File: internal/ingest/ingest.go
// Package ingest baut den Vektorindex aus den BookStack-Inhalten auf.
package ingest

import (
	"context"
	"fmt"
	"strings"

	"wiki-rog/internal/bookstack"
	"wiki-rog/internal/chunk"
	"wiki-rog/internal/config"
	"wiki-rog/internal/ollama"
	"wiki-rog/internal/store"
)

type Stats struct {
	Pages          int
	Chunks         int
	CollectionSize int
}

func pageHeader(p bookstack.Page) string {
	var parts []string
	for _, s := range []string{p.BookName, p.ChapterName, p.Name} {
		if strings.TrimSpace(s) != "" {
			parts = append(parts, s)
		}
	}
	return strings.Join(parts, " > ")
}

// Run liest BookStack aus, zerlegt/embedded die Seiten und schreibt sie in den Store.
// reset=true leert den Store zuvor komplett.
func Run(ctx context.Context, cfg config.Config, st store.Store, progress func(string)) (Stats, error) {
	if err := cfg.RequireBookStack(); err != nil {
		return Stats{}, err
	}
	oc := ollama.New(cfg.OllamaURL)
	if err := oc.Check(ctx); err != nil {
		return Stats{}, err
	}
	bs := bookstack.New(cfg.BookStackURL, cfg.BookStackTokenID, cfg.BookStackTokenSecret, cfg.BookStackVerifySSL)

	var stats Stats
	err := bs.IterPages(cfg.BookIDs, cfg.ShelfIDs, func(p bookstack.Page) error {
		// Alte Chunks dieser Seite entfernen (inkrementelles Update).
		if err := st.DeletePage(ctx, p.ID); err != nil {
			return err
		}
		header := pageHeader(p)
		chunks := chunk.Text(p.Content, cfg.ChunkSize, cfg.ChunkOverlap)
		if len(chunks) == 0 {
			return nil
		}
		items := make([]store.Item, 0, len(chunks))
		for idx, c := range chunks {
			embedInput := c
			if header != "" {
				embedInput = header + "\n\n" + c
			}
			emb, err := oc.Embed(ctx, cfg.OllamaEmbedModel, embedInput)
			if err != nil {
				return err
			}
			items = append(items, store.Item{
				ID:        fmt.Sprintf("page-%d-chunk-%d", p.ID, idx),
				Embedding: emb,
				Document:  c,
				Meta: store.Meta{
					PageID:      p.ID,
					PageName:    p.Name,
					BookName:    p.BookName,
					ChapterName: p.ChapterName,
					URL:         p.URL,
					ChunkIndex:  idx,
					UpdatedAt:   p.UpdatedAt,
				},
			})
		}
		if err := st.Upsert(ctx, items); err != nil {
			return err
		}
		stats.Pages++
		stats.Chunks += len(chunks)
		if progress != nil {
			progress(fmt.Sprintf("indexiert: %s (%d Chunks)", p.Name, len(chunks)))
		}
		return nil
	})
	if err != nil {
		return stats, err
	}
	size, err := st.Count(ctx)
	if err != nil {
		return stats, err
	}
	stats.CollectionSize = size
	return stats, nil
}

File: src/ingest.py
"""Ingest-Pipeline: BookStack -> Chunks -> Embeddings -> Vektorspeicher."""
from __future__ import annotations

from tqdm import tqdm

from config import Config
from src.bookstack_client import BookStackClient, Page
from src.chunker import chunk_text
from src.ollama_client import OllamaClient
from src.vectorstore import VectorStore


def _page_header(page: Page) -> str:
    """Kontextzeile, die jedem Chunk vorangestellt wird (bessere Retrieval-Treffer)."""
    parts = [p for p in [page.book_name, page.chapter_name, page.name] if p]
    return " > ".join(parts)


def ingest(cfg: Config, reset: bool = False) -> dict:
    cfg.require_bookstack()

    bs = BookStackClient(
        cfg.bookstack_url, cfg.bookstack_token_id, cfg.bookstack_token_secret
    )
    ollama = OllamaClient(cfg.ollama_url)
    ollama.check()

    store = VectorStore(cfg.chroma_dir, cfg.chroma_collection)
    if reset:
        store.reset()

    pages = 0
    chunks_total = 0

    page_iter = bs.iter_pages(book_ids=cfg.book_ids, shelf_ids=cfg.shelf_ids)
    for page in tqdm(page_iter, desc="Seiten indexieren", unit="Seite"):
        # Bei erneutem Lauf ohne reset: alte Chunks dieser Seite entfernen
        if not reset:
            store.delete_page(page.id)

        header = _page_header(page)
        chunks = chunk_text(page.content, cfg.chunk_size, cfg.chunk_overlap)
        if not chunks:
            continue

        ids, embeddings, documents, metadatas = [], [], [], []
        for idx, chunk in enumerate(chunks):
            embed_input = f"{header}\n\n{chunk}" if header else chunk
            embeddings.append(ollama.embed(cfg.ollama_embed_model, embed_input))
            ids.append(f"page-{page.id}-chunk-{idx}")
            documents.append(chunk)
            metadatas.append(
                {
                    "page_id": page.id,
                    "page_name": page.name,
                    "book_name": page.book_name,
                    "chapter_name": page.chapter_name,
                    "url": page.url,
                    "chunk_index": idx,
                    "updated_at": page.updated_at,
                }
            )

        store.add(ids, embeddings, documents, metadatas)
        pages += 1
        chunks_total += len(chunks)

    return {
        "pages": pages,
        "chunks": chunks_total,
        "collection_size": store.count(),
    }
